File: 05-ejemplos/javascript/01_basicos.js
// 01 - Variables, tipos, operadores, conversiones y control de flujo.
// Ejecutar:  node 01_basicos.js

// ---------------------------------------------------------------- VARIABLES
const nombre = "Ana";        // no se puede reasignar (úsala por defecto)
let edad = 25;               // se puede reasignar
// var  → NO usar (ámbito de función, hoisting confuso)

console.log(typeof nombre, typeof edad, typeof true, typeof undefined, typeof null, typeof {}, typeof []);
// string number boolean undefined object object object   (null y [] dicen "object": rarezas de JS)
console.log(Array.isArray([]));                            // true (la forma correcta de detectar arrays)

// ---------------------------------------------------------------- OPERADORES
console.log(7 + 2, 7 - 2, 7 * 2, 7 / 2, 7 % 2, 2 ** 10);    // 9 5 14 3.5 1 1024
console.log(Math.floor(7 / 2), Math.trunc(-3.5), Math.round(2.5), Math.round(-2.5)); // 3 -3 3 -2
console.log(0.1 + 0.2, (0.1 + 0.2).toFixed(2), 0.1 + 0.2 === 0.3); // 0.30000000000000004 "0.30" false

// Igualdad: SIEMPRE ===
console.log("5" == 5, "5" === 5, null == undefined, null === undefined, NaN === NaN); // true false true false false

// Truthy / falsy: son falsy → false, 0, "", null, undefined, NaN
console.log(Boolean(""), Boolean("0"), Boolean([]), Boolean(0), !!"hola");  // false true true false true

// Lógicos que devuelven valores
const usuario = null;
console.log(usuario || "Invitado");      // "Invitado"  (|| usa el derecho si el izquierdo es falsy)
console.log(0 || 10, 0 ?? 10);           // 10 0        (?? solo si es null/undefined)
const config = { tema: { color: "azul" } };
console.log(config?.tema?.color, config?.fuente?.tamano);   // azul undefined (sin error)

// ---------------------------------------------------------------- CONVERSIONES
console.log(Number("42"), Number(""), Number("abc"), parseInt("42px"), parseFloat("3.5kg")); // 42 0 NaN 42 3.5
console.log(String(10) + 1, 10 + 1 + "1", "3" * "4", "3" + 4);   // "101" "111" 12 "34"
console.log(Number.isNaN(Number("abc")), Number.isInteger(5.0), (1234.5678).toFixed(2));
console.log((1234567.891).toLocaleString("es-EC", { style: "currency", currency: "USD" }));

// ---------------------------------------------------------------- TEMPLATE STRINGS
const precio = 12.5;
console.log(`Hola ${nombre}, en 5 años tendrás ${edad + 5}. Total: $${(precio * 3).toFixed(2)}`);
const multilinea = `Línea 1
Línea 2`;
console.log(multilinea);

// ---------------------------------------------------------------- CONDICIONALES
const nota = 8.4;
if (nota >= 9) console.log("Excelente");
else if (nota >= 7) console.log("Aprobado");
else console.log("Reprobado");

console.log(edad >= 18 ? "Mayor" : "Menor");

const dia = 3;
switch (dia) {
  case 1:
  case 7:
    console.log("Fin de semana");
    break;
  case 2: case 3: case 4: case 5: case 6:
    console.log("Laborable");
    break;                     // ¡sin break sigue al siguiente case!
  default:
    console.log("Inválido");
}

// ---------------------------------------------------------------- BUCLES
for (let i = 0; i < 3; i++) console.log("for", i);

const frutas = ["pera", "uva", "kiwi"];
for (const f of frutas) console.log("of", f);             // valores (arrays, strings)
for (const i in frutas) console.log("in", i);             // índices/claves (mejor para objetos)
for (const [i, f] of frutas.entries()) console.log(i, f);

let n = 5, suma = 0;
while (n > 0) { suma += n; n--; }
console.log("suma", suma);

let intentos = 0;
do { intentos++; } while (intentos < 3);
console.log("intentos", intentos);

// break y continue
for (const x of [3, 8, 11, 4]) {
  if (x % 2 === 0) continue;
  if (x > 10) { console.log("encontré", x); break; }
}

// Etiquetas para romper bucles anidados
externo: for (let i = 0; i < 3; i++) {
  for (let j = 0; j < 3; j++) {
    if (i * j === 2) { console.log("rompo en", i, j); break externo; }
  }
}

// ---------------------------------------------------------------- ENTRADA POR CONSOLA (Node)
// import readline from "readline/promises";   (en archivos .mjs o con "type": "module")
// const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
// const texto = await rl.question("Número: ");
// rl.close();
// Argumentos:  node 01_basicos.js 5 hola   →  process.argv = [node, archivo, "5", "hola"]
console.log("argumentos:", process.argv.slice(2));

File: 05-ejemplos/python/01_basicos.py
"""01 - Variables, tipos, operadores, entrada/salida y control de flujo.
Ejecutar:  python 01_basicos.py
"""

# ---------------------------------------------------------------- VARIABLES Y TIPOS
nombre = "Ana"          # str
edad = 25               # int
estatura = 1.65         # float
activo = True           # bool
nada = None             # NoneType

print(type(nombre), type(edad), type(estatura), type(activo), type(nada))
# <class 'str'> <class 'int'> <class 'float'> <class 'bool'> <class 'NoneType'>

# Asignación múltiple e intercambio
a, b = 1, 2
a, b = b, a
print("a =", a, "b =", b)                     # a = 2 b = 1

# ---------------------------------------------------------------- OPERADORES
print(7 + 2, 7 - 2, 7 * 2, 7 / 2)             # 9 5 14 3.5   (/ siempre da float)
print(7 // 2, 7 % 2, 2 ** 10)                 # 3 1 1024     (división entera, residuo, potencia)
print(-7 // 2)                                # -4  (redondea hacia abajo, ¡ojo!)
print(round(3.14159, 2), abs(-5), max(3, 9, 1), min(3, 9, 1))   # 3.14 5 9 1

# Comparación y lógicos
print(5 > 3 and 2 > 1, 5 > 3 or 2 > 5, not True)   # True True False
print(1 < edad < 100)                               # True  (comparación encadenada)
print(0.1 + 0.2 == 0.3, abs(0.1 + 0.2 - 0.3) < 1e-9)  # False True  (¡floats imprecisos!)

# ---------------------------------------------------------------- CONVERSIONES
print(int("42") + 1, float("3.5") * 2, str(10) + "0", bool(""), bool("hola"))
# 43 7.0 100 False True
print(int(3.99), int("  7  "))               # 3 7  (int trunca; strip automático de espacios)

# ---------------------------------------------------------------- FORMATO DE SALIDA
precio = 1234.5
print(f"Hola {nombre}, tienes {edad} años")          # Hola Ana, tienes 25 años
print(f"Precio: ${precio:,.2f}")                     # Precio: $1,234.50
print(f"{'Izq':<10}|{'Der':>10}|{'Centro':^10}|")    # alinear columnas
print(f"{7:03d}", f"{0.256:.1%}")                    # 007 25.6%
print(f"{edad=}")                                    # edad=25  (útil para depurar)
print("a", "b", "c", sep="-", end="!\n")             # a-b-c!

# ---------------------------------------------------------------- IF / ELIF / ELSE
nota = 8.4
if nota >= 9:
    estado = "Excelente"
elif nota >= 7:
    estado = "Aprobado"
else:
    estado = "Reprobado"
print(estado)                                        # Aprobado

# Ternario
print("Mayor" if edad >= 18 else "Menor")            # Mayor

# match (Python 3.10+) — como switch
dia = 3
match dia:
    case 1 | 7:
        print("Fin de semana")
    case 2 | 3 | 4 | 5 | 6:
        print("Día laborable")                       # Día laborable
    case _:
        print("Inválido")

# ---------------------------------------------------------------- BUCLES
for i in range(3):                    # 0,1,2
    print("for", i)
for i in range(10, 0, -3):            # 10,7,4,1
    print(i, end=" ")
print()

total = 0
n = 5
while n > 0:
    total += n
    n -= 1
print("Suma 1..5 =", total)           # 15

# break, continue, else en bucles
for x in [3, 8, 11, 4]:
    if x % 2 == 0:
        continue                      # salta los pares
    if x > 10:
        print("Encontré", x)          # Encontré 11
        break
else:
    print("No se ejecuta porque hubo break")

# Tabla de multiplicar con bucles anidados
for i in range(1, 4):
    print(" ".join(f"{i * j:3d}" for j in range(1, 6)))

# ---------------------------------------------------------------- ENTRADA (descomenta para probar)
# numero = int(input("Ingrese un número: "))
# print("El doble es", numero * 2)

# Validar entrada en bucle
# while True:
#     texto = input("Edad: ")
#     if texto.isdigit():
#         edad = int(texto)
#         break
#     print("Debe ser un número entero")
